From 495a52ffc2202ef07fd80c0727a0c504250ddaae Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Thu, 20 Mar 2025 10:03:00 +0100
Subject: [PATCH 0193/1010] net/page_pool: dovetail: allow for high order pages
 in oob cache

At this chance, fix a page leak at init since we weren't releasing the
page returned by __page_pool_alloc_pages_slow() when populating the
oob cache.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/net/page_pool.h |   8 +--
 net/core/page_pool.c    | 105 ++++++++++++++++++++++++++++++++--------
 2 files changed, 86 insertions(+), 27 deletions(-)

diff --git a/include/net/page_pool.h b/include/net/page_pool.h
index 777633fee0..43052bfeb2 100644
--- a/include/net/page_pool.h
+++ b/include/net/page_pool.h
@@ -268,15 +268,9 @@ static inline bool page_pool_is_oob(struct page_pool *pool)
 
 static inline unsigned int page_pool_cache_size(struct page_pool *pool)
 {
-	return IS_ENABLED(CONFIG_PAGE_POOL_OOB) &&
-		pool->p.flags & PP_FLAG_PAGE_OOB ? pool->p.pool_size : PP_ALLOC_CACHE_SIZE;
+	return page_pool_is_oob(pool) ? pool->p.pool_size : PP_ALLOC_CACHE_SIZE;
 }
 
-static inline unsigned int page_pool_cache_refill(struct page_pool *pool)
-{
-	return IS_ENABLED(CONFIG_PAGE_POOL_OOB) &&
-		pool->p.flags & PP_FLAG_PAGE_OOB ? pool->p.pool_size : PP_ALLOC_CACHE_REFILL;
-}
 #else
 static inline void page_pool_destroy(struct page_pool *pool)
 {
diff --git a/net/core/page_pool.c b/net/core/page_pool.c
index 57aaee2265..641e23e24e 100644
--- a/net/core/page_pool.c
+++ b/net/core/page_pool.c
@@ -156,9 +156,7 @@ static void page_pool_producer_unlock(struct page_pool *pool,
 		spin_unlock_bh(&pool->ring.producer_lock);
 }
 
-noinline
-static struct page *__page_pool_alloc_pages_slow(struct page_pool *pool,
-						gfp_t gfp);
+static int page_pool_init_cache(struct page_pool *pool);
 
 static int page_pool_init(struct page_pool *pool,
 			  const struct page_pool_params *params)
@@ -236,22 +234,13 @@ static int page_pool_init(struct page_pool *pool,
 	if (pool->p.flags & PP_FLAG_DMA_MAP)
 		get_device(pool->p.dev);
 
-#ifdef CONFIG_PAGE_POOL_OOB
-	pool->alloc.cache = kzalloc_node(
-		sizeof(struct page *) * params->pool_size,
-		GFP_KERNEL, params->nid);
-	if (pool->alloc.cache == NULL) {
+	if (page_pool_init_cache(pool)) {
 #ifdef CONFIG_PAGE_POOL_STATS
 		free_percpu(pool->recycle_stats);
 #endif
 		ptr_ring_cleanup(&pool->ring, NULL);
 		return -ENOMEM;
 	}
-	raw_spin_lock_init(&pool->alloc.oob_lock);
-	/* Populate the fast cache of oob-accessed pools at init. */
-	if (page_pool_is_oob(pool))
-		__page_pool_alloc_pages_slow(pool, GFP_KERNEL);
-#endif
 
 	return 0;
 }
@@ -285,6 +274,9 @@ static struct page *page_pool_refill_alloc_cache(struct page_pool *pool)
 	struct page *page;
 	int pref_nid; /* preferred NUMA node */
 
+	if (WARN_ON_ONCE(page_pool_is_oob(pool)))
+		return NULL;
+
 	/* Quicker fallback, avoid locks when ring is empty */
 	if (__ptr_ring_empty(r)) {
 		alloc_stat_inc(pool, empty);
@@ -320,7 +312,7 @@ static struct page *page_pool_refill_alloc_cache(struct page_pool *pool)
 			page = NULL;
 			break;
 		}
-	} while (pool->alloc.count < page_pool_cache_refill(pool));
+	} while (pool->alloc.count < PP_ALLOC_CACHE_REFILL);
 
 	/* Return last page */
 	if (likely(pool->alloc.count > 0)) {
@@ -426,12 +418,83 @@ static struct page *__page_pool_alloc_page_order(struct page_pool *pool,
 	return page;
 }
 
+static void page_pool_return_page(struct page_pool *pool, struct page *page);
+
+static int page_pool_init_cache(struct page_pool *pool)
+{
+#ifdef CONFIG_PAGE_POOL_OOB
+	unsigned int pool_size = pool->p.pool_size;
+	unsigned int pp_order = pool->p.order;
+	unsigned int pp_flags = pool->p.flags;
+	int i, nr_pages, nid = pool->p.nid;
+	struct page *page;
+
+	pool->alloc.cache = kzalloc_node(sizeof(struct page *) * pool_size,
+					GFP_KERNEL, nid);
+	if (pool->alloc.cache == NULL)
+		return -ENOMEM;
+
+	raw_spin_lock_init(&pool->alloc.oob_lock);
+
+	if (!page_pool_is_oob(pool))
+		return 0;
+
+	/*
+	 * Populate the fast cache of an oob-accessed pool at
+	 * init. Unlike for regular in-band pools, high order pages
+	 * may live in the fast cache of oob pools.
+	 */
+	if (unlikely(pp_order)) {
+		for (nr_pages = 0; nr_pages < pool_size; nr_pages++) {
+			page = __page_pool_alloc_page_order(pool, GFP_KERNEL);
+			if (!page)
+				goto fail;
+			pool->alloc.cache[nr_pages] = page;
+		}
+	} else {
+		nr_pages = alloc_pages_bulk_array_node(
+			GFP_KERNEL, nid, pool_size, pool->alloc.cache);
+		if (unlikely(nr_pages < pool_size))
+			goto fail;
+
+		for (i = 0; i < nr_pages; i++) {
+			page = pool->alloc.cache[i];
+			if ((pp_flags & PP_FLAG_DMA_MAP) &&
+				unlikely(!page_pool_dma_map(pool, page)))
+				goto fail;
+
+			page_pool_set_pp_info(pool, page);
+			pool->alloc.cache[i] = page;
+			pool->pages_state_hold_cnt++;
+			trace_page_pool_state_hold(pool, page,
+						pool->pages_state_hold_cnt);
+		}
+	}
+
+	pool->alloc.count = nr_pages;
+
+	return 0;
+
+fail:
+	while (nr_pages-- > 0) {
+		page = pool->alloc.cache[nr_pages];
+		page_pool_return_page(pool, page);
+	}
+
+	kfree(pool->alloc.cache);
+
+	return -ENOMEM;
+#endif	/* !CONFIG_PAGE_POOL_OOB */
+
+	return 0;
+}
+
 /* slow path */
 noinline
 static struct page *__page_pool_alloc_pages_slow(struct page_pool *pool,
 						 gfp_t gfp)
 {
-	const int bulk = page_pool_cache_refill(pool);
+	const int bulk = PP_ALLOC_CACHE_REFILL;
 	unsigned int pp_flags = pool->p.flags;
 	unsigned int pp_order = pool->p.order;
 	struct page *page;
@@ -646,9 +709,11 @@ __page_pool_put_page(struct page_pool *pool, struct page *page,
 		/*
 		 * If a page from an oob-accessed pool is still
 		 * referenced when released or the fast cache would
-		 * overflow, we bark at the console then bluntly leak
-		 * it. This situation should never occur, or would
-		 * denote a serious API usage issue anyway.
+		 * unexpectedly overflow (meaning we attempt to double
+		 * release or free an invalid page), we bark at the
+		 * console then bluntly leak it. This situation should
+		 * never occur, or would denote a critical bug in the
+		 * calling code.
 		 */
 		if (WARN_ON(dovetail_debug() && page_ref_count(page) != 1))
 			return NULL;
@@ -658,10 +723,10 @@ __page_pool_put_page(struct page_pool *pool, struct page *page,
 		raw_spin_lock_irqsave(&pool->alloc.oob_lock, flags);
 		ret = page_pool_recycle_in_cache(page, pool);
 		raw_spin_unlock_irqrestore(&pool->alloc.oob_lock, flags);
-		WARN_ON(!ret);
+		WARN_ON(dovetail_debug() && !ret);
 		return NULL;
 	}
-#endif
+#endif	/* !CONFIG_PAGE_POOL_OOB */
 
 	/* This allocator is optimized for the XDP mode that uses
 	 * one-frame-per-page, but have fallbacks that act like the
-- 
2.30.2

